FutureFeatures.md

```diff
@@ -66,16 +66,160 @@ Options under consideration:
+would effectively be skipped.
+
+Another important issue is mapping WebIDL values types that aren't simple
+[primitive types](http://www.w3.org/TR/WebIDL/#dfn-primitive-type):
+* [Dictionary types](http://www.w3.org/TR/WebIDL/#idl-dictionary)
+  would [appear](http://www.w3.org/TR/WebIDL/#es-dictionary) to require
+  JS objects but are actually defined as values such that they can
+  be (and are, in various browser implementations) flattened to C structs.
+  Thus, a natural WebAssembly binding would be to map dictionaries to structs
+  in linear memory passed by reference (integer offset).
+* The same goes for [sequence types](http://www.w3.org/TR/WebIDL/#idl-sequence).
+* [Enumeration types](http://www.w3.org/TR/WebIDL/#es-enumeration) could be
+  mapped to canonical integers.
+* [Union types](http://www.w3.org/TR/WebIDL/#idl-union) could be handled in
+  multiple ways. One option is to treat the union type itself as an importable
+  opaque reference type (when all the elements are themselves reference types).
+  Another option is to introduce an overload of each signature for each element
+  of the union type such that all calls passed a single element type and the
+  full Union Type was never explicitly represented in WebAssembly.
+* [Callback function types](http://www.w3.org/TR/WebIDL/#es-callback-function)
+  could map to a `(function pointer, environment pointer)` closure pair.
+
+Overall, the goal of mapping WebIDL to WebAssembly builtin modules is to avoid
+the need to define a duplicate WebAssembly interface for all Web APIs.  In
+practice, some WebIDL patterns may have an unnatural or inefficient mapping
+into WebAssembly such that new overloads and best practices would need to be
+adopted. Over time, though, these rough edges would be ironed out leaving the
+long term benefit of defining Web APIs with a single interface and ensuring
+that JS and WebAssembly always had access to the same raw functionality.
+
+### Direct GC access
+
+In contract to *opaque* reference types, a second sub-feature would be to allow
+direct GC allocation and field access from WebAssembly code through
+*non-opaque* reference types.
+
+There is a lot of the design left to
+consider for this feature, but a few points of tentative agreement are:
+* To avoid baking in a single language's object model, define low-level GC
+  primitives (viz., structs and arrays) and allow the source language compiler
+  to build up features like virtual dispatch and access control.
+* GC struct and array types would have associated *struct/array reference
+  types* that were similar to and symmetric with 
+  [opaque reference types](FutureFeatures.md#opaque-reference-types)
+  (just not opaque).
+* The GC heap would be semantically distinct from linear memory and thus
+  the fields of GC objects could safely hold reference types (unlike linear
+  memory). GC object fields could also hold 
+  [function pointers](AstSemantics.md#calls) directly (again, unlike
```

Comment:
I think we need a better way to phrase this, but I'm having trouble coming up with something...

I don't think you ever want the canonical representation in the heap to be a function pointer - you need type information, etc. But you could definitely have a representation that is (function type info, function pointer) and mechanisms for calling into a stored fp with an expected signature. This would mean a type check, but it would predict well.

If the GC object fields store a function of a *specific signature*, we can enforce that signature all the way through the system and that should reduce an invocation into a raw function pointer call because we don't need to type-check at call time, just at write time.